src/diffusion/model.py

```diff
@@ -64,3 +88,92 @@ def forward(
         ).last_hidden_state
 
         return output
+
+    def training_step(self, batch, batch_idx):
+        context, gt = batch
+
+        emb = self.emb(gt)
+
+        x_0, x_t, t = get_diffusion_variables(self.diffusion_steps, emb, self.alphas_cumprod_prev)
+
+        x_0_hat = self(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_t,
+            time_ids=t,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+        probs = self.classifier(x_0)
+        ce = calculate_batch_ce(probs, gt, pad_mask)
+
+        emb_mask = ~pad_mask.unsqueeze(-1)
+
+        mse = torch.where(
+            t == 1,
+            flat_mean((x_0_hat - emb) ** 2 * emb_mask),
+            flat_mean((x_0_hat - x_0) ** 2 * emb_mask),
+        ).mean()
+
+        t0_loss = (x_0**2 * emb_mask).mean()
+
+        loss = mse + ce + t0_loss
+
+        metrics = {
+            "train/mse": mse.item(),
+            "train/ce": ce.item(),
+            "train/t0": t0_loss.item(),
+            "train/loss": loss.item(),
+        }
```

Comment:
Lightning recommend to avoid `.item()` calls (https://pytorch-lightning.readthedocs.io/en/stable/guides/speed.html#things-to-avoid). Just pass tensors here:
```python
losses = {"train/mse": mse, "train/ce": ce, "train/t0": t0_loss, "train/loss": loss }
```